### PR TITLE
Implementation by Pavel Drankov

### PR DESCRIPTION
**Pull request by Sammers21**
My implementation contains some changes which might not be related to the tasks, adding `.editorconfig` file or changes in the `.gitignore` file, but I think they are fine. 

I'm also thinking that using rxJava2 is fine. If it is not, just let me know.

About git history... it does not consist of 3 commits(so one for each task), but from more(it is 10 now!). 

There are 3 from them, and each of them is related to a task:

1. c530d94a9c83275012e59bf4ef06206c8d8942f2 relates to `Setup and basic retrieval` 
2. 162cb98 relates to `Post new quotes` 
3. b890263 relates to `Realtime quotes`

In any case, if you think that git history is not organized well enough, just let me know and I will reorganize it in a way so the PR will contain only 3 commits.

### CODE REVIEW

**Comment by tsegismont** [comment]
@Sammers21 I merged the other two PRs, can you please rebase?

**Comment by Sammers21** [comment]
@tsegismont , sure, done.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 1]
```diff
@@ -32,19 +51,28 @@ public void start(Future<Void> startFuture) throws Exception {
...
     Future<Void> future = Future.future();
-    jdbcClient.getConnection(getConn -> {
-      if (getConn.succeeded()) {
-        SQLConnection connection = getConn.result();
-        connection.execute("RUNSCRIPT FROM '" + script + "'", exec -> {
-          connection.close();
-          if (exec.succeeded()) {
-            future.succeeded();
+
+    vertx.fileSystem().readFile(script, fileReadResult -> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: no doubt this works but why did you remove the `RUNSCRIPT` version? it's much more concise and really enough for our use case. just curious

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 2]
```diff
@@ -25,6 +29,38 @@ public void start(Future<Void> startFuture) throws Exception {
...
+          });
+        } else {
+          logger.error("failed in db connection establishing", getConn.cause());
+          responseWithError(response);
+        }
+      });
+    });
+
+    int port = config().getInteger("http.port", 8080);
+    server.requestHandler(router::accept).listen(port);
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: The server may fail to bind and the verticle would still be considered deployed successfully. You should use the `listen` method with a handler argument.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 3]
```diff
@@ -42,11 +79,11 @@ public void start(Future<Void> startFuture) throws Exception {
...
       }
     });
+  }
 
-    int port = config().getInteger("http.port", 8080);
-
-    // TODO setup database and web server and eventually...
-    startFuture.complete();
+  private void responseWithError(HttpServerResponse response) {
+    response.setStatusCode(400);
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Does not look like the proper status code for this kind of error :-)

File: src/main/resources/db.sql [hunk 4]
```diff
@@ -1,5 +1,6 @@
+DROP TABLE IF EXISTS quotes;
```
Comment (tsegismont) [inline src/main/resources/db.sql:1]: Why do you need this? I have used such a script in the past, and since the H2 db is in-memory, it should be created from scratch when the server starts

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 5]
```diff
@@ -25,6 +29,38 @@ public void start(Future<Void> startFuture) throws Exception {
 
     jdbcClient = JDBCClient.createShared(vertx, jdbcConfig);
 
+    // web server setup
+    HttpServer server = vertx.createHttpServer();
+    Router router = Router.router(vertx);
+
+    router.route("/quotes").handler(routingContext -> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: minor but `router.get` would be more specfic

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 6]
```diff
@@ -33,7 +35,47 @@ public void start(Future<Void> startFuture) throws Exception {
...
+        String author = request.getString("author");
+        if (author == null) {
+          author = DEFAULT_AUTHOR_VALUE;
+        }
+
+        String finalAuthor = author;
+        jdbcClient.getConnection(getConn -> {
+          if (getConn.succeeded()) {
+            SQLConnection connection = getConn.result();
+            String sqlToExecute = String.format("INSERT INTO quotes (text,author) VALUES ('%s', '%s');", text, finalAuthor);
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: this is subject to sql injection attacks. it's better to user `connection.updateWithParams`

File: src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java [hunk 7]
```diff
@@ -41,9 +43,33 @@ public void setup(TestContext testContext) {
...
+      ws.handler(data -> {
+        JsonObject json = data.toJsonObject();
+        testContext.assertEquals(json.getString("author"), SAMPLE_AUTHOR_NAME);
+        testContext.assertEquals(json.getString("text"), SAMPLE_QUOTE_TEXT);
+        async.countDown();
+      });
+    });
+
+    // wait some time for websocket connection establishing and subscription availability
+    Thread.sleep(1000);
```
Comment (tsegismont) [inline src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java]: Try to eliminate this by using `Async`

**Comment by tsegismont** [comment]
@Sammers21 good job!

Your proposal for streaming new quotes is interesting. I was thinking about using the eventbus (we could imagine that messages come from other nodes in cluster mode). But RxJava subject is fine for a simple standalone solution

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: oh, sorry, I simply not familiar with `RUNSCRIPT` command and I thought that the command is like `TODO` and should be replaced

File: src/main/resources/db.sql (hunk 4 above)
Comment (Sammers21) [reply src/main/resources/db.sql:1]: I faced some problems when running tests without such addition. I might guess, that during test process H2 created only once, so not created each time for each test case.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: which one do you think should be used?

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 5 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Yes, agree, it was changed to `router.get` in https://github.com/tsegismont/vertx-gsoc-2018-qotd/commit/162cb98ac70c1730f75b9e633dcdf08bad3b7cca when implementing POST /quotes

**Comment by Sammers21** [comment]
@tsegismont, I was also thinking about using event-bus, I used it when implementing it for the first time. 

The reason why I am not using event bus here is simple: let's imagine two consumers who consuming messages from the same address. When a message arrives, only one of them will receive the message.

This also means that when we will have two WebSocket consumers on `/realtime`, only one of them will receive a message when update happen.

But, we also can use a combined approach and have only one consumer on an event bus address. When a message arrived to publish it to dbUpdatesPublisher, so we will benefit from using evet bus and all WebSocket listeners will receive their updates.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 6 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Agree

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Agree, done.

File: src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java (hunk 7 above)
Comment (Sammers21) [reply src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java]: Done

**Comment by tsegismont** [comment]
> When a message arrives, only one of them will receive the message.

Not if use `eventBus.publish()` instead of `eventBus.send()`

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 5 above)
Comment (tsegismont) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: :+1:

File: src/main/resources/db.sql (hunk 4 above)
Comment (tsegismont) [reply src/main/resources/db.sql:1]: Can you tell me how to reproduce?

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 3 above)
Comment (tsegismont) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Well this is invoked when server gets in trouble so I'll let you propose something which is appropriate (in the 5xx range). Definitely not a `400 BAD REQUEST`

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 8]
```diff
@@ -31,6 +30,13 @@
 
   @Override
   public void start(Future<Void> startFuture) {
+    Future<Void> dbReady = Future.future();
+    Future<Void> httpServerReady = Future.future();
+
+    // complete startFuture only when db and http server started
+    compose(startFuture, dbReady, httpServerReady);
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Check out the Vert.x core documentation, in particular the _sequential composition_ section

File: src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java [hunk 9]
```diff
@@ -61,11 +61,10 @@ public void testRealtimeWebSocket(TestContext testContext) throws InterruptedExc
       });
     });
 
-    // wait some time for websocket connection establishing and subscription availability
-    Thread.sleep(1000);
-
     // force receiving a update
-    postQuote(SAMPLE_AUTHOR_NAME, SAMPLE_QUOTE_TEXT, testContext);
+    for (int i = 0; i < 100; i++) {
```
Comment (tsegismont) [inline src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java]: This will work most often but is not bullet proof. Try to find a way to start posting when the websocket has been established

Comment (tsegismont) [reply src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java]: ha, you did it in the next commit :+1:

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 8 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Done, replaced with `CompositeFuture.all`

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 10]
```diff
@@ -79,7 +78,7 @@ public void start(Future<Void> startFuture) {
                 JsonObject newQuote = new JsonObject()
                   .put(AUTHOR_FILED, finalAuthor)
                   .put(TEXT_FILED, text);
-                dbUpdatesPublisher.onNext(newQuote);
+                vertx.eventBus().publish(DB_UPDATES_ADDRESS, newQuote);
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: :+1:

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 11]
```diff
@@ -32,9 +33,13 @@ public void start(Future<Void> startFuture) {
     Future<Void> dbReady = Future.future();
     Future<Void> httpServerReady = Future.future();
 
-    // complete startFuture only when db and http server started
-    compose(startFuture, dbReady, httpServerReady);
-    compose(startFuture, httpServerReady, dbReady);
+    CompositeFuture.all(dbReady, httpServerReady).setHandler(result -> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: This is concurrent composition. We don't want everything to be started  concurrently, we want these steps, in order:

1. setup the schema
2. import base data
3. start the http server.

In your example, the http server could start reply requests without the DB being ready

Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Ah, got it.

Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Just changed, so now the steps are done sequentially

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Agree

File: src/main/resources/db.sql (hunk 4 above)
Comment (Sammers21) [reply src/main/resources/db.sql:1]: You need to have at least two tests which use QuoteOfTheDayVerticle.

Comment (Sammers21) [reply src/main/resources/db.sql:1]: When deploying the verticle for the first time, it will create an instance of h2, which will be reused for the second test, so you will get something like this:
```
Table "quotes" already exists; SQL statement:
CREATE TABLE quotes (
  quote_id BIGINT AUTO_INCREMENT PRIMARY KEY NOT NULL,
  text VARCHAR NOT NULL,
  author VARCHAR NOT NULL
);
 [42101-196]

	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.command.ddl.CreateTable.update(CreateTable.java:117)
	at org.h2.command.CommandContainer.update(CommandContainer.java:101)
	at org.h2.command.Command.executeUpdate(Command.java:260)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:192)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:164)
	at com.mchange.v2.c3p0.impl.NewProxyStatement.execute(NewProxyStatement.java:75)
	at io.vertx.ext.jdbc.impl.actions.JDBCExecute.execute(JDBCExecute.java:46)
	at io.vertx.ext.jdbc.impl.actions.JDBCExecute.execute(JDBCExecute.java:31)
	at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.handle(AbstractJDBCAction.java:58)
	at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.lambda$execute$0(AbstractJDBCAction.java:66)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

Comment (Sammers21) [reply src/main/resources/db.sql:1]: The fastest way to reproduce is to either create a second test in the skeleton repository or simply:
1. clone my fork
2. remove `DROP TABLE IF EXISTS quotes;` line in the `db.sql` file
3. try to run tests

Comment (tsegismont) [reply src/main/resources/db.sql:1]: @Sammers21 ha I see. Ok then that's fine.

**Comment by tsegismont** [comment]
LGTM :+1:
